### DISCUSSION TITLE
How do custom scripts work with Hammerdb via the CLI only?

### DISCUSSION THREAD

**Discussion by Hamz4h** [TPC-Council/HammerDB#520] [General]
Hey - I'm exploring HammerDB for a particular simulation use case however I'm using a Mac and therefore have settled on the docker container variant and the CLI to try and put together a simple use case. I've been able to wire everything up with postgres and run the predefined script. Everything works wonderfully.

The bit I'm not understanding is `customscript` - I want to add a new call to a stored proc in postgres - this stored proc is as basic as can be for now. The way I assumed this would work via the CLI is I would be able to edit the default script provided by `loadscript`. This doesn't seem possible via the CLI and the CLI blog post references opening the script in the GUI and editing it there.

What is the correct way for me to edit the tcl script for postgres (in this case)? Is this even possible or have I misunderstood how far the editing capability goes? Does it stop at the values marked as `Editable`? I.e. in the tcl file, I'm unable to add a new `proc` which runs a custom `pg_exec` to call the stored proc?

**Comment by sm-shaw** [comment]
Hi, great question, No you haven't misunderstood, you can edit anything and even write and load your own scripts if you wish.  The reason these have to be in Tcl is explained [here](https://www.hammerdb.com/blog/uncategorized/why-tcl-is-700-faster-than-python-for-database-benchmarking/) so we get GIL free parallel execution, so any valid Tcl syntax is fine - and of course all of this can be extended in C as well so the scope is unlimited. 

So, yes the intended process is you use the GUI editor to make your template script, save it and then load it in the CLI. On a MAC you have the alternatives of running Windows or Linux under virtualbox to get a GUI, however I'll also include a command later on to write out the script to a file directly from the CLI. 

As an example of how it works, here is the most basic test script we can use (for placeholding the first character on the first line should be #) 

![test1](https://user-images.githubusercontent.com/38044085/222689026-de763f2b-1834-42b6-be70-19443d2ed32f.png)

So we run it and the green tick shows it worked. 

![test3](https://user-images.githubusercontent.com/38044085/222689172-c5380fec-6996-47a7-98f9-89bd05cda53f.png)

So now we go to the file menu and save it

![test2](https://user-images.githubusercontent.com/38044085/222689376-bf26d8ac-369a-4c6a-9266-570a9f6cc7e1.png)

We look at the file, and we get what was in the GUI (of course in this case we could have typed it all directly if we wished) 
```
$ cat pgtest.tcl 
#!/usr/local/bin/tclsh8.6
puts "This is a test script"
```
So now we use customscript to load our new script and run it, and you can see we get the same output. 
```
HammerDB-4.7$ ./hammerdbcli 
HammerDB CLI v4.7
Copyright (C) 2003-2023 Steve Shaw
Type "help" for a list of commands
Initialized SQLite on-disk database /tmp/hammer.DB using existing tables (36,864 KB)
hammerdb>customscript /opt/pgtest.tcl
Loaded /opt/pgtest.tcl

hammerdb>vurun
Vuser 1 created - WAIT IDLE
1 Virtual Users Created
Vuser 1:RUNNING
Vuser 1:This is a test script
Vuser 1:FINISHED SUCCESS
ALL VIRTUAL USERS COMPLETE
jobid=6401BE275F5F03E233237393
hammerdb>
```
So far so good, so now as another example we want to run a PostgreSQL driver script but only run the read only transactions so just slev and ostat. This time we have edited the script to comment out the other stored procedures and saved it in the same way. 

![roedit](https://user-images.githubusercontent.com/38044085/222690442-62119f10-3b75-4b1e-b967-5c0d48e37659.png)

Note that as we are using a customscript HammerDB doesn't know if it's a timed test (needing an extra monitor virtual user) or a test one (without) - so we need to manually set the number of virtual users we need, in this case 2, one to monitor and one to run the workload.  
We run it and get no NOPM as we only ran slev and ostat. 

```
hammerdb>vuset vu 2

hammerdb>customscript /opt/rotest.tcl
Loaded /opt/rotest.tcl

hammerdb>vurun
Vuser 1 created - WAIT IDLE
Vuser 2 created - WAIT IDLE
2 Virtual Users Created
Vuser 1:RUNNING
Vuser 1:Beginning rampup time of 0 minutes
Vuser 1:Rampup complete, Taking start Transaction Count.
Vuser 1:Timing test period of 1 in minutes
Vuser 2:RUNNING
Vuser 2:Processing 10000000 transactions with output suppressed...
Vuser 1:1 ...,
Vuser 1:Test complete, Taking end Transaction Count.
Vuser 1:1 Active Virtual Users configured
Vuser 1:TEST RESULT : System achieved 0 NOPM from 105781 PostgreSQL TPM
Vuser 1:FINISHED SUCCESS
Vuser 2:FINISHED SUCCESS
ALL VIRTUAL USERS COMPLETE
jobid=6401BFC85F5F03E293830393
hammerdb>
```

It should be clear that you can modify the script to do absolutely anything that you want. e.g. there are examples in the documentation on how to test over HTTP as well https://www.hammerdb.com/docs/ch15s01.html.  The script then gets sent to the Virtual Users to be executed in parallel by them. 

That is the recommend approach, as the GUI will fill out all the options for you to make the template script to work from. 

However, of course anything with HammerDB is possible, so if you only have the CLI and you've done a loadscript and can see the contents with 'print script' then the following command will write it out to a file for you:

`hammerdb>puts [open "/opt/mypgfile.tcl" w ] $_ED(package)`

You can then edit it with a text editor and load it with customscript to run as you wish.

**Reply by Hamz4h** [reply]
Awesome! I shall give this a go and report back but it everything makes sense Steve, thank you!!

**Reply by Hamz4h** [reply]
Hey Steve, so the crux of my problem was this statement:

> Note that as we are using a customscript HammerDB doesn't know if it's a timed test (needing an extra monitor virtual user) or a test one (without) - so we need to manually set the number of virtual users we need, in this case 2, one to monitor and one to run the workload.

My config for testing was to set up a single vu with my custom script. After changing this to 2 users, no issues, everything worked as I had wanted it to.

With the above write-up, however, I'm much more confident in the approach. Thanks again Steve, massively appreciated!

**Comment by sm-shaw** [comment]
This is a high value discussion topic, as it shows that the ability to edit and run your own scripts is being used in hammerDB and highlighted an example where the CLI can be used, but the GUI can't. 

For this reason it may not be a good idea to unify the GUI and CLI scripts as described here https://github.com/TPC-Council/HammerDB/issues/385 and leave the interface as it is with the script editor showing and editing what the virtual users actually run. 

Instead, it now appears to make more sense to add functionality to make editing and running your own scripts in the CLI easier: 

1. A save script command. As we use in other places, this would find a suitable temp location and write out the contents you see in "print script" for manual editing and reloading. 
2. Add an option to the customscript script command to set a script you have loaded to timed or test so if it is timed then a monitor virtual user is created as you would expect.

**Comment by sm-shaw** [comment]
PR #522 adds this functionality
